1-Draft/RFCNNNN-ML-Based-Command-Line-Completion.md

```diff
@@ -0,0 +1,174 @@
+---
+RFC: RFC
+Author: Paul Higinbotham, Travis Plunk, Dongbo Wang
+Status: Draft
+SupercededBy: N/A
+Version: 1.0
+Area: Engine
+Comments Due: March 30, 2018
+---
+
+# Machine Learning based Command Line Completion Suggestions for PowerShell
+
+The RFC proposes enhancing PowerShell's existing command line completion with machine learning based suggestions.  
+
+## Motivation
+
+PowerShell's existing command line completion looks for closest matches of known commands, cmdlets, and cmdlet parameters. The match is performed on the text fragment currently typed on the command line. Commands are shown alphabetically and parameters based on parameter set.  
+
+It would be interesting if machine learning could enhance this by providing completion results based on common usage patterns. These patterns could be from the global community that would help new users of PowerShell. Or they could be based on the current user's history to help that user recall complex patterns without having to refer to notes. The machine learning model could provide a richer experience by encompassing context over multiple commands either piped together or typed on multiple lines.  
+
+At this point it is not clear what the right machine learning model is to provide the best experience. It is expected to take many iterations of models and data processing before good and useful suggestions are provided. Customer privacy is critically important and so this feature will be opt-in since data collection is required to train the models. Collected data may be kept locally but more likely will be transferred to the cloud to create a larger data set for training. We must adhere to General Data Protection Regulation (GDPR) requirements to ensure that data is safe and in control of the customer.  
+
+The architecture will allow running components locally or in the cloud.  The cloud has tremendous processing power for training and running models, but latency may make locally running components more attractive in some cases.  
+
+Machine Learning is an inherently experimental process where different models and data features are examined, trained and evaluated. This RFC does not attempt to solve the machine learning and user experience problems, but instead provide an architecture to support experimentation of various Machine learning models and UI candidates.  
+
+### Goals
+
+* Support for both local and cloud based implementations of any component
+* Integrates with existing PowerShell command completion
+* Provides for privacy protection
+* Supports feature opt-in
+
+## Specification 
+
+### Modules
+#### Prediction Module  
+* Collects command line text used in PowerShell for model training
+* Returns prediction from submitted command line fragment
+* Creates suggestion success statistics based on user choice
+
+#### General Data Protection Regulation (GDPR) support Module

```

Comment:
Is this module the telemetry module? If so, maybe we can change the name of this section.

Reply:
Yes, we can call this the telemetry module that supports GDPR.